ethcore/src/engines/authority_round/mod.rs

```diff
@@ -450,20 +450,29 @@ impl Engine<EthereumMachine> for AuthorityRound {
 	}
 
 	fn seals_internally(&self) -> Option<bool> {
+		// TODO: accept a `&Call` here so we can query the validator set.
 		Some(self.signer.read().is_some())
 	}
 
 	/// Attempt to seal the block internally.
 	///
 	/// This operation is synchronous and may (quite reasonably) not be available, in which case
 	/// `Seal::None` will be returned.
-	fn generate_seal(&self, block: &ExecutedBlock) -> Seal {
+	fn generate_seal(&self, block: &ExecutedBlock, parent: &Header) -> Seal {
 		// first check to avoid generating signature most of the time
 		// (but there's still a race to the `compare_and_swap`)
 		if !self.can_propose.load(AtomicOrdering::SeqCst) { return Seal::None; }
 
 		let header = block.header();
+		let parent_step: U256 = header_step(parent)
+			.expect("Header has been verified; qed").into();
+
 		let step = self.step.load();
+		let expected_diff = U256::from(U128::max_value()) + parent_step - step.into();
```

Comment:
Looks like the difficulty calcution is already copy pasted in couple of places, maybe we should extract it into a helper function?